File: native/cocos/scripting/js-bindings/manual/jsb_videoplayer.cpp
/****************************************************************************
 Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
 worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
 not use Cocos Creator software for developing other software or tools that's
 used for developing games. You are not granted to publish, distribute,
 sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

#include "jsb_videoplayer.hpp"

#include "cocos/scripting/js-bindings/jswrapper/SeApi.h"
#include "cocos/scripting/js-bindings/manual/jsb_conversions.hpp"
#include "cocos/scripting/js-bindings/manual/jsb_global.h"

#include "base/ccUTF8.h"
#include "platform/CCApplication.h"
#include "ui/videoplayer/VideoPlayer.h"

using namespace cocos2d;
using namespace cocos2d::network;

/*
 [Constructor(in DOMString url, in optional DOMString protocols)]
 [Constructor(in DOMString url, in optional DOMString[] protocols)]
 interface VideoPlayer {
 readonly attribute DOMString url;

 // ready state
 const unsigned short CONNECTING = 0;
 const unsigned short OPEN = 1;
 const unsigned short CLOSING = 2;
 const unsigned short CLOSED = 3;
 readonly attribute unsigned short readyState;
 readonly attribute unsigned long bufferedAmount;

 // networking
 attribute Function onopen;
 attribute Function onmessage;
 attribute Function onerror;
 attribute Function onclose;
 readonly attribute DOMString protocol;
 void send(in DOMString data);
 void close();
 };
 VideoPlayer implements EventTarget;
 */

se::Class *__jsb_VideoPlayer_class = nullptr;

static bool VideoPlayer_finalize(se::State &s)
{
    VideoPlayer *cobj = (VideoPlayer *)s.nativeThisObject();
    CCLOGINFO("jsbindings: finalizing JS object %p (VideoPlayer)", cobj);

    cobj->stop();
    cobj->autorelease();

    return true;
}
SE_BIND_FINALIZE_FUNC(VideoPlayer_finalize)

static bool VideoPlayer_constructor(se::State &s)
{
    se::Object *obj = s.thisObject();
    VideoPlayer *cobj = new (std::nothrow) VideoPlayer();
    obj->setPrivateData(cobj);

    return true;
}
SE_BIND_CTOR(VideoPlayer_constructor, __jsb_VideoPlayer_class, VideoPlayer_finalize)

static bool VideoPlayer_close(se::State &s)
{
    VideoPlayer *cobj = (VideoPlayer *)s.nativeThisObject();

    // Attach current VideoPlayer instance to global object to prevent VideoPlayer instance
    // being garbage collected after "ws.close(); ws = null;"
    // There is a state that current VideoPlayer JS instance is being garbaged but its finalize
    // callback has not be invoked. Then in "JSB_VideoPlayerDelegate::onClose", se::Object is
    // still be able to be found and while invoking JS 'onclose' method, crash will happen since
    // JS instance is invalid and is going to be collected. This bug is easiler reproduced on iOS
    // because JavaScriptCore is more GC sensitive.
    // Please note that we need to detach it from global object in "JSB_VideoPlayerDelegate::onClose".
    se::ScriptEngine::getInstance()->getGlobalObject()->attachObject(s.thisObject());
    return true;
}
SE_BIND_FUNC(VideoPlayer_close)

static bool VideoPlayer_addEventListener(se::State& s)
{
    VideoPlayer* cobj = (VideoPlayer*)s.nativeThisObject();
    SE_PRECONDITION2(cobj, false, "VideoPlayer_addEventListener : Invalid Native Object");
    const auto& args = s.args();
    size_t argc = args.size();
    CC_UNUSED bool ok = true;
    if (argc == 2) {
        std::string arg0;
        ok &= seval_to_std_string(args[0], &arg0);
        
        std::function<void ()> arg1;
        do {
            if (args[1].isObject() && args[1].toObject()->isFunction())
            {
                se::Value jsThis(s.thisObject());
                se::Value jsFunc(args[1]);
                jsThis.toObject()->attachObject(jsFunc.toObject());
                auto lambda = [=]() -> void {
                    se::ScriptEngine::getInstance()->clearException();
                    se::AutoHandleScope hs;
                    
                    CC_UNUSED bool ok = true;
                    se::ValueArray args;
                    se::Value rval;
                    se::Object* thisObj = jsThis.isObject() ? jsThis.toObject() : nullptr;
                    se::Object* funcObj = jsFunc.toObject();
                    bool succeed = funcObj->call(args, thisObj, &rval);
                    if (!succeed) {
                        se::ScriptEngine::getInstance()->clearException();
                    }
                };
                arg1 = lambda;
            }
            else
            {
                arg1 = []() -> void {};
            }
        } while(false)
        ;
        SE_PRECONDITION2(ok, false, "js_cocos2dx_experimental_video_VideoPlayer_addEventListener : Error processing arguments");
        cobj->addEventListener(arg0, arg1);
        return true;
    }
    SE_REPORT_ERROR("wrong number of arguments: %d, was expecting %d", (int)argc, 1);
    return false;
}
SE_BIND_FUNC(VideoPlayer_addEventListener)

static bool VideoPlayer_removeEventListener(se::State &s)
{
    VideoPlayer* cobj = (VideoPlayer*)s.nativeThisObject();
    SE_PRECONDITION2(cobj, false, "VideoPlayer_addEventListener : Invalid Native Object");
    const auto& args = s.args();
    size_t argc = args.size();
    CC_UNUSED bool ok = true;
    if (argc == 2) {
        std::string arg0;
        ok &= seval_to_std_string(args[0], &arg0);
        std::function<void ()> arg1 = []() -> void {};
        SE_PRECONDITION2(ok, false, "VideoPlayer_addEventListener : Error processing arguments");
        cobj->addEventListener(arg0, arg1);
        return true;
    }
    SE_REPORT_ERROR("wrong number of arguments: %d, was expecting %d", (int)argc, 1);
    return false;
}
SE_BIND_FUNC(VideoPlayer_removeEventListener)

static bool VideoPlayer_setURL(se::State &s)
{
    VideoPlayer* cobj = (VideoPlayer*)s.nativeThisObject();
    SE_PRECONDITION2(cobj, false, "VideoPlayer_setURL : Invalid Native Object");
    const auto& args = s.args();
    size_t argc = args.size();
    CC_UNUSED bool ok = true;
    if (argc == 1) {
        std::string arg0;
        ok &= seval_to_std_string(args[0], &arg0);
        SE_PRECONDITION2(ok, false, "VideoPlayer_setURL : Error processing arguments");
        cobj->setURL(arg0);
        return true;
    }
    SE_REPORT_ERROR("wrong number of arguments: %d, was expecting %d", (int)argc, 1);
    return false;
}
SE_BIND_FUNC(VideoPlayer_setURL)

static bool VideoPlayer_play(se::State &s)
{
    VideoPlayer* cobj = (VideoPlayer*)s.nativeThisObject();
    SE_PRECONDITION2(cobj, false, "VideoPlayer_play : Invalid Native Object");
    const auto& args = s.args();
    size_t argc = args.size();
    if (argc == 0) {
        cobj->play();
        return true;
    }
    SE_REPORT_ERROR("wrong number of arguments: %d, was expecting %d", (int)argc, 0);
    return false;
}
SE_BIND_FUNC(VideoPlayer_play)

static bool VideoPlayer_pause(se::State &s)
{
    VideoPlayer* cobj = (VideoPlayer*)s.nativeThisObject();
    SE_PRECONDITION2(cobj, false, "VideoPlayer_play : Invalid Native Object");
    const auto& args = s.args();
    size_t argc = args.size();
    if (argc == 0) {
        cobj->pause();
        return true;
    }
    SE_REPORT_ERROR("wrong number of arguments: %d, was expecting %d", (int)argc, 0);
    return false;
}
SE_BIND_FUNC(VideoPlayer_pause)

static bool VideoPlayer_stop(se::State &s)
{
    VideoPlayer* cobj = (VideoPlayer*)s.nativeThisObject();
    SE_PRECONDITION2(cobj, false, "js_cocos2dx_experimental_video_VideoPlayer_stop : Invalid Native Object");
    const auto& args = s.args();
    size_t argc = args.size();
    if (argc == 0) {
        cobj->stop();
        return true;
    }
    SE_REPORT_ERROR("wrong number of arguments: %d, was expecting %d", (int)argc, 0);
    return false;
}
SE_BIND_FUNC(VideoPlayer_stop)

static bool VideoPlayer_seekTo(se::State &s)
{
    VideoPlayer* cobj = (VideoPlayer*)s.nativeThisObject();
    SE_PRECONDITION2(cobj, false, "VideoPlayer_seekTo : Invalid Native Object");
    const auto& args = s.args();
    size_t argc = args.size();
    CC_UNUSED bool ok = true;
    if (argc == 1) {
        float arg0 = 0;
        ok &= seval_to_float(args[0], &arg0);
        SE_PRECONDITION2(ok, false, "VideoPlayer_seekTo : Error processing arguments");
        cobj->seekTo(arg0);
        return true;
    }
    SE_REPORT_ERROR("wrong number of arguments: %d, was expecting %d", (int)argc, 1);
    return false;
}
SE_BIND_FUNC(VideoPlayer_seekTo)

static bool VideoPlayer_currentTime(se::State &s)
{
    VideoPlayer* cobj = (VideoPlayer*)s.nativeThisObject();
    SE_PRECONDITION2(cobj, false, "VideoPlayer_currentTime : Invalid Native Object");
    const auto& args = s.args();
    size_t argc = args.size();
    CC_UNUSED bool ok = true;
    if (argc == 0) {
        float result = cobj->currentTime();
        ok &= float_to_seval(result, &s.rval());
        SE_PRECONDITION2(ok, false, "VideoPlayer_currentTime : Error processing arguments");
        return true;
    }
    SE_REPORT_ERROR("wrong number of arguments: %d, was expecting %d", (int)argc, 0);
    return false;
}
SE_BIND_FUNC(VideoPlayer_currentTime)

static bool VideoPlayer_duration(se::State &s)
{
    VideoPlayer* cobj = (VideoPlayer*)s.nativeThisObject();
    SE_PRECONDITION2(cobj, false, "VideoPlayer_duration : Invalid Native Object");
    const auto& args = s.args();
    size_t argc = args.size();
    CC_UNUSED bool ok = true;
    if (argc == 0) {
        float result = cobj->duration();
        ok &= float_to_seval(result, &s.rval());
        SE_PRECONDITION2(ok, false, "VideoPlayer_duration : Error processing arguments");
        return true;
    }
    SE_REPORT_ERROR("wrong number of arguments: %d, was expecting %d", (int)argc, 0);
    return false;
}
SE_BIND_FUNC(VideoPlayer_duration)

static bool VideoPlayer_setKeepAspectRatioEnabled(se::State &s)
{
    VideoPlayer* cobj = (VideoPlayer*)s.nativeThisObject();
    SE_PRECONDITION2(cobj, false, "VideoPlayer_setKeepAspectRatioEnabled : Invalid Native Object");
    const auto& args = s.args();
    size_t argc = args.size();
    CC_UNUSED bool ok = true;
    if (argc == 1) {
        bool arg0;
        ok &= seval_to_boolean(args[0], &arg0);
        SE_PRECONDITION2(ok, false, "VideoPlayer_setKeepAspectRatioEnabled : Error processing arguments");
        cobj->setKeepAspectRatioEnabled(arg0);
        return true;
    }
    SE_REPORT_ERROR("wrong number of arguments: %d, was expecting %d", (int)argc, 1);
    return false;
}
SE_BIND_FUNC(VideoPlayer_setKeepAspectRatioEnabled)

static bool VideoPlayer_isKeepAspectRatioEnabled(se::State &s)
{
    VideoPlayer* cobj = (VideoPlayer*)s.nativeThisObject();
    SE_PRECONDITION2(cobj, false, "VideoPlayer_onPlayEvent : Invalid Native Object");
    const auto& args = s.args();
    size_t argc = args.size();
    CC_UNUSED bool ok = true;
    if (argc == 0) {
        bool result = cobj->isKeepAspectRatioEnabled();
        ok &= boolean_to_seval(result, &s.rval());
        SE_PRECONDITION2(ok, false, "VideoPlayer_currentTime : Error processing arguments");
        return true;
    }
    SE_REPORT_ERROR("wrong number of arguments: %d, was expecting %d", (int)argc, 1);
    return false;
}
SE_BIND_FUNC(VideoPlayer_isKeepAspectRatioEnabled)

static bool VideoPlayer_setFullScreenEnabled(se::State &s)
{
    VideoPlayer* cobj = (VideoPlayer*)s.nativeThisObject();
    SE_PRECONDITION2(cobj, false, "VideoPlayer_setFullScreenEnabled : Invalid Native Object");
    const auto& args = s.args();
    size_t argc = args.size();
    CC_UNUSED bool ok = true;
    if (argc == 1) {
        bool arg0;
        ok &= seval_to_boolean(args[0], &arg0);
        SE_PRECONDITION2(ok, false, "VideoPlayer_setFullScreenEnabled : Error processing arguments");
        cobj->setFullScreenEnabled(arg0);
        return true;
    }
    SE_REPORT_ERROR("wrong number of arguments: %d, was expecting %d", (int)argc, 1);
    return false;
}
SE_BIND_FUNC(VideoPlayer_setFullScreenEnabled)

static bool VideoPlayer_setVisible(se::State &s)
{
    VideoPlayer* cobj = (VideoPlayer*)s.nativeThisObject();
    SE_PRECONDITION2(cobj, false, "VideoPlayer_setVisible : Invalid Native Object");
    const auto& args = s.args();
    size_t argc = args.size();
    CC_UNUSED bool ok = true;
    if (argc == 1) {
        bool arg0;
        ok &= seval_to_boolean(args[0], &arg0);
        SE_PRECONDITION2(ok, false, "VideoPlayer_setVisible : Error processing arguments");
        cobj->setVisible(arg0);
        return true;
    }
    SE_REPORT_ERROR("wrong number of arguments: %d, was expecting %d", (int)argc, 1);
    return false;
}
SE_BIND_FUNC(VideoPlayer_setVisible)

static bool VideoPlayer_setFrame(se::State &s)
{
    VideoPlayer* cobj = (VideoPlayer*)s.nativeThisObject();
    SE_PRECONDITION2(cobj, false, "VideoPlayer_setViewPos : Invalid Native Object");
    auto& args = s.args();
    size_t argc = args.size();
    CC_UNUSED bool ok = true;
    if (argc == 4) {
        float x = 0;
        ok = seval_to_float(args[0], &x);
        SE_PRECONDITION2(ok, false, "Convert arg0 failed!");
        float y = 0;
        ok = seval_to_float(args[1], &y);
        SE_PRECONDITION2(ok, false, "Convert arg1 failed!");
        int32_t width = 0;
        ok = seval_to_int32(args[2], &width);
        SE_PRECONDITION2(ok, false, "Convert arg2 failed!");
        int32_t height = 0;
        ok = seval_to_int32(args[3], &height);
        SE_PRECONDITION2(ok, false, "Convert arg3 failed!");
        cobj->setFrame(x, y, width, height);
        return true;
    }
    SE_REPORT_ERROR("wrong number of arguments: %d, was expecting %d", (int)argc, 4);
    return false;
}
SE_BIND_FUNC(VideoPlayer_setFrame)

static bool VideoPlayer_setViewSize(se::State &s)
{
    return true;
}
SE_BIND_FUNC(VideoPlayer_setViewSize)

bool register_all_videoplayer(se::Object *obj)
{
    se::Class *cls = se::Class::create("VideoPlayerCore", obj, nullptr, _SE(VideoPlayer_constructor));
    cls->defineFinalizeFunction(_SE(VideoPlayer_finalize));
    cls->defineFunction("close", _SE(VideoPlayer_close));
    cls->defineFunction("setURL", _SE(VideoPlayer_setURL));
    cls->defineFunction("play", _SE(VideoPlayer_play));
    cls->defineFunction("pause", _SE(VideoPlayer_pause));
    cls->defineFunction("stop", _SE(VideoPlayer_stop));
    cls->defineFunction("seekTo", _SE(VideoPlayer_seekTo));
    cls->defineFunction("currentTime", _SE(VideoPlayer_currentTime));
    cls->defineFunction("duration", _SE(VideoPlayer_duration));
    cls->defineFunction("setKeepAspectRatioEnabled", _SE(VideoPlayer_setKeepAspectRatioEnabled));
    cls->defineFunction("isKeepAspectRatioEnabled", _SE(VideoPlayer_isKeepAspectRatioEnabled));
    cls->defineFunction("setFullScreenEnabled", _SE(VideoPlayer_setFullScreenEnabled));
    cls->defineFunction("addEventListener", _SE(VideoPlayer_addEventListener));
    cls->defineFunction("removeEventListener", _SE(VideoPlayer_removeEventListener));
    cls->defineFunction("setVisible", _SE(VideoPlayer_setVisible));
    cls->defineFunction("setViewSize", _SE(VideoPlayer_setViewSize));
    cls->defineFunction("setFrame", _SE(VideoPlayer_setFrame));
    cls->install();

    JSBClassType::registerClass<VideoPlayer>(cls);

    __jsb_VideoPlayer_class = cls;

    se::ScriptEngine::getInstance()->clearException();

    return true;
}

File: port/native/engine/jsb-videoplayer.js
/****************************************************************************
 Copyright (c) 2018 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
  worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
  not use Cocos Creator software for developing other software or tools that's
  used for developing games. You are not granted to publish, distribute,
  sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

(function () {
    if (!(cc && cc.VideoPlayer && cc.VideoPlayer.Impl)) {
        return;
    }

    var vec3 = cc.Vec3;
    var _worldMat = new cc.Mat4();

    var _topLeft = new vec3();
    var _bottomRight = new vec3();

    var _impl = cc.VideoPlayer.Impl;
    var _p = cc.VideoPlayer.Impl.prototype;

    _p._bindEvent = function () {
        let video = this._video,
            self = this;

        if (!video) {
            return;
        }

        //binding event
        let cbs = this.__eventListeners;
        cbs.loadedmetadata = function () {
            self._loadedmeta = true;
            self._dispatchEvent(_impl.EventType.META_LOADED);
        };
        cbs.ended = function () {
            if (self._video !== video) return;
            self._playing = false;
            self._dispatchEvent(_impl.EventType.COMPLETED);
        };
        cbs.play = function () {
            if (self._video !== video) return;
            self._playing = true;
            self._dispatchEvent(_impl.EventType.PLAYING);
        };
        cbs.pause = function () {
            if (self._ignorePause || self._video !== video) return;
            self._playing = false;
            self._dispatchEvent(_impl.EventType.PAUSED);
        };
        cbs.click = function () {
            self._dispatchEvent(_impl.EventType.CLICKED);
        };
        cbs.stoped = function () {
            self._dispatchEvent(_impl.EventType.STOPPED);
            self._ignorePause = false;
        };

        video.addEventListener("loadedmetadata", cbs.loadedmetadata);
        video.addEventListener("ended", cbs.ended);
        video.addEventListener("play", cbs.play);
        video.addEventListener("pause", cbs.pause);
        video.addEventListener("click", cbs.click);
        video.addEventListener("stoped", cbs.stoped);

        function onCanPlay() {
            if (this._loaded)
                return;

            this._loaded = true;
            this._dispatchEvent(_impl.EventType.READY_TO_PLAY);
            this._updateVisibility();
        }

        cbs.onCanPlay = onCanPlay.bind(this);
        video.addEventListener('canplay', cbs.onCanPlay);
        video.addEventListener('canplaythrough', cbs.onCanPlay);
        video.addEventListener('suspend', cbs.onCanPlay);
    };

    _p._updateVisibility = function () {
        if (!this._video) return;
        let video = this._video;
        if (this._visible) {
            this._video.setVisible(true);
        } else {
            this._video.setVisible(false);
            video.pause();
            this._playing = false;
        }
    };

    _p._updateSize = function (width, height) {

    };

    _p.createDomElementIfNeeded = function () {
        if (!jsb.VideoPlayer) {
            cc.warn('VideoPlayer is not supported.');
            return null;
        }

        if (!this._video) {
            this._video = new jsb.VideoPlayer();
        }
    };

    _p.removeDom = function () {
        let video = this._video;
        if (video) {
            video.stop()
            video.setVisible(false)

            let cbs = this.__eventListeners;

            cbs.loadedmetadata = null;
            cbs.ended = null;
            cbs.play = null;
            cbs.pause = null;
            cbs.click = null;
            cbs.onCanPlay = null;

            video.destroy();
        }

        this._video = null;
        this._url = "";
    };

    _p.setURL = function (path) {
        let source, extname;

        if (this._url === path) {
            return;
        }

        this.removeDom();

        this._url = path;
        this.createDomElementIfNeeded();
        this._bindEvent();

        let video = this._video;
        if (!video) {
            return;
        }

        video.setVisible(this._visible);

        this._loaded = false;
        this._played = false;
        this._playing = false;
        this._loadedmeta = false;

        video.setURL(this._url);
        this._forceUpdate = true;
    };

    _p.getURL = function() {
        return this._url;
    };

    _p.play = function () {
        let video = this._video;
        if (!video || !this._visible || this._playing) return;

        video.play();
        this._playing = true;
    };

    _p.setStayOnBottom = function (enabled) {};

    _p.pause = function () {
        let video = this._video;
        if (!this._playing || !video) return;

        video.pause();
        this._playing = false;
    };

    _p.resume = function () {
        let video = this._video;
        if (this._playing || !video) return;

        video.resume();
        this._playing = true;
    };

    _p.stop = function () {
        let video = this._video;
        if (!video || !this._visible) return;
        this._ignorePause = true;

        video.stop();
        this._playing = false;
    };

    _p.setVolume = function (volume) {

    };


    _p.seekTo = function (time) {
        let video = this._video;
        if (!video) return;

        if (this._loaded) {
            video.seekTo(time)
        } else {
            let cb = function () {
                video.seekTo(time)
            };
            video.addEventListener(_impl._polyfill.event, cb);
        }
        if (_impl._polyfill.autoplayAfterOperation && this.isPlaying()) {
            setTimeout(function () {
                video.play();
            }, 20);
        }
    };

    _p.isPlaying = function () {
        return this._playing;
    };

    _p.duration = function () {
        let video = this._video;
        let duration = -1;
        if (!video) return duration;

        duration = video.duration();
        if (duration <= 0) {
            cc.logID(7702);
        }

        return duration;
    };


    _p.currentTime = function () {
        let video = this._video;
        if (!video) return -1;

        return video.currentTime();
    };

    _p.setKeepAspectRatioEnabled = function (isEnabled) {
        if (!this._video) {
            return false;
        }
        return this._video.setKeepAspectRatioEnabled(isEnabled);
    };

    _p.isKeepAspectRatioEnabled = function () {
        if (!this._video) {
            return false;
        }
        return this._video.isKeepAspectRatioEnabled();
    };

    _p.isFullScreenEnabled = function () {
        return this._fullScreenEnabled;
    };

    _p.setEventListener = function (event, callback) {
        this._EventList[event] = callback;
    };

    _p.removeEventListener = function (event) {
        this._EventList[event] = null;
    };

    _p._dispatchEvent = function (event) {
        let callback = this._EventList[event];
        if (callback)
            callback.call(this, this, this._video.src);
    };

    _p.onPlayEvent = function () {
        let callback = this._EventList[_impl.EventType.PLAYING];
        callback.call(this, this, this._video.src);
    };

    _p.enable = function () {
        let list = _impl.elements;
        if (list.indexOf(this) === -1)
            list.push(this);
        this.setVisible(true);
    };

    _p.disable = function () {
        let list = _impl.elements;
        let index = list.indexOf(this);
        if (index !== -1)
            list.splice(index, 1);
        this.setVisible(false);
    };

    _p.destroy = function () {
        this.disable();
        this.removeDom();
    };

    _p.setVisible = function (visible) {
        if (this._visible !== visible) {
            this._visible = !!visible;
            this._updateVisibility();
        }
    };

    _p.setFullScreenEnabled = function (enable) {
        let video = this._video;
        if (!video) {
            return;
        }
        this._fullScreenEnabled = enable;
        video.setFullScreenEnabled(enable);
    };

    _p.updateMatrix = function (node) {
        if (!this._video || !this._visible) return;

        node.getWorldMatrix(_worldMat);
        if (!this._forceUpdate &&
            this._m00 === _worldMat.m[0] && this._m01 === _worldMat.m[1] &&
            this._m04 === _worldMat.m[4] && this._m05 === _worldMat.m[5] &&
            this._m12 === _worldMat.m[12] && this._m13 === _worldMat.m[13] &&
            this._w === node._contentSize.width && this._h === node._contentSize.height) {
            return;
        }

        // update matrix cache
        this._m00 = _worldMat.m[0];
        this._m01 = _worldMat.m[1];
        this._m04 = _worldMat.m[4];
        this._m05 = _worldMat.m[5];
        this._m12 = _worldMat.m[12];
        this._m13 = _worldMat.m[13];
        this._w = node._contentSize.width;
        this._h = node._contentSize.height;

        let camera = cc.Camera.findCamera(node)._camera;

        let canvas_width = cc.game.canvas.width;
        let canvas_height = cc.game.canvas.height;
        let ap = node._anchorPoint;
        // Vectors in node space
        vec3.set(_topLeft, - ap.x * this._w, (1.0 - ap.y) * this._h, 0);
        vec3.set(_bottomRight, (1 - ap.x) * this._w, - ap.y * this._h, 0);
        // Convert to world space
        vec3.transformMat4(_topLeft, _topLeft, _worldMat);
        vec3.transformMat4(_bottomRight, _bottomRight, _worldMat);
        // Convert to Screen space
        camera.worldToScreen(_topLeft, _topLeft, canvas_width, canvas_height);
        camera.worldToScreen(_bottomRight, _bottomRight, canvas_width, canvas_height);

        let finalWidth = _bottomRight.x - _topLeft.x;
        let finalHeight = _topLeft.y - _bottomRight.y;
        this._video.setFrame(_topLeft.x, canvas_height - _topLeft.y, finalWidth, finalHeight);
    };

    _impl.EventType = {
        PLAYING: 0,
        PAUSED: 1,
        STOPPED: 2,
        COMPLETED: 3,
        META_LOADED: 4,
        CLICKED: 5,
        READY_TO_PLAY: 6
    };

    // video 队列，所有 vidoe 在 onEnter 的时候都会插入这个队列
    _impl.elements = [];
    // video 在 game_hide 事件中被自动暂停的队列，用于回复的时候重新开始播放
    _impl.pauseElements = [];

    cc.game.on(cc.game.EVENT_HIDE, function () {
        let list = _impl.elements;
        for (let element, i = 0; i < list.length; i++) {
            element = list[i];
            if (element.isPlaying()) {
                element.pause();
                _impl.pauseElements.push(element);
            }
        }
    });

    cc.game.on(cc.game.EVENT_SHOW, function () {
        let list = _impl.pauseElements;
        let element = list.pop();
        while (element) {
            element.play();
            element = list.pop();
        }
    });
})();
